### PR TITLE
Add MAV_COMPONENT documentation

### PR DESCRIPTION
**Pull request by hamishwillee**
As part of general push to get all common.xml entities documented, this pushes some proposed text for MAV_COMPONENT. The descriptions should cover how and when each value is used, and any exceptions. Ideally we should attempt to restrict use and intent so it is clear how it should be used and extended, and remove ambiguity that might later result in different uses. 

I have put in a bunch of placeholders for discussion. Please highlight anything you think should be more precise, or could cause potential issue.

### CODE REVIEW

File: message_definitions/v1.0/common.xml [hunk 1]
```diff
@@ -369,131 +369,132 @@
...
-        <description/>
+        <description>Servo ID #12.</description>
       </entry>
       <entry value="152" name="MAV_COMP_ID_SERVO13">
-        <description/>
+        <description>Servo ID #13.</description>
       </entry>
       <entry value="153" name="MAV_COMP_ID_SERVO14">
-        <description/>
+        <description>Servo ID #14.</description>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml]: Just FYI, if there will ever be more servos, it is inconvenient that the next value has been used.

File: message_definitions/v1.0/common.xml [hunk 2]
```diff
@@ -369,131 +369,132 @@
...
-        <description/>
+        <description>Gimbal ID for QX1.</description>
       </entry>
       <entry value="160" name="MAV_COMP_ID_FLARM">
-        <description/>
+        <description>ID for FLARM collision alert component.</description>
       </entry>
       <entry value="180" name="MAV_COMP_ID_MAPPER">
-        <description/>
+        <description></description>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml]: What is this for?

**Comment by hamishwillee** [comment]
@auturgy @WickedShell FYI.

File: message_definitions/v1.0/common.xml [hunk 3]
```diff
@@ -369,131 +369,132 @@
...
       </entry>
       <entry value="190" name="MAV_COMP_ID_MISSIONPLANNER">
-        <description/>
+        <description>ID for a component that supports the Mission microservice.</description>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml]: In enum description I suggest components should use the most appropriate value, which is fine for cameras, servos, whatever. But what should an autopilot do for a case like this? It supports missions, but is not separately addressed for that purpose. You could similarly consider this case for a system that supports obstacle avoidance etc.

File: message_definitions/v1.0/common.xml [hunk 4]
```diff
@@ -369,131 +369,132 @@
...
       </entry>
       <entry value="195" name="MAV_COMP_ID_PATHPLANNER">
-        <description/>
+        <description></description>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml]: @mrivi @baumanta How would you characterise a PATHPLANNER component for someone who did not know of your work? - "It is a component that ..." Presumably this is for the companion side code so you can address just a particular node with your messages?

File: message_definitions/v1.0/common.xml [hunk 5]
```diff
@@ -369,131 +369,132 @@
...
       </entry>
       <entry value="196" name="MAV_COMP_ID_OBSTACLE_AVOIDANCE">
-        <description/>
+        <description></description>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml]: @mrivi @baumanta Ditto for MAV_COMP_ID_OBSTACLE_AVOIDANCE

File: message_definitions/v1.0/common.xml [hunk 6]
```diff
@@ -369,131 +369,132 @@
...
-        <description/>
+        <description>GPS #1.</description>
       </entry>
       <entry value="221" name="MAV_COMP_ID_GPS2">
-        <description/>
+        <description>GPS #2.</description>
       </entry>
       <entry value="240" name="MAV_COMP_ID_UDP_BRIDGE">
-        <description/>
+        <description></description>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml]: A bridge connects two transports - is a UDP bridge TO UDP, and what is it from. Same question for MAV_COMP_ID_UART_BRIDGE below

File: message_definitions/v1.0/common.xml [hunk 7]
```diff
@@ -369,131 +369,132 @@
...
       </entry>
       <entry value="241" name="MAV_COMP_ID_UART_BRIDGE">
-        <description/>
+        <description></description>
       </entry>
       <entry value="250" name="MAV_COMP_ID_SYSTEM_CONTROL">
-        <description/>
+        <description></description>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml]: What does this component do?

File: message_definitions/v1.0/common.xml [hunk 8]
```diff
@@ -369,131 +369,132 @@
...
-        <description/>
+        <description>Servo ID #14.</description>
       </entry>
       <entry value="154" name="MAV_COMP_ID_GIMBAL">
-        <description/>
+        <description>Generic Gimbal ID. Use ID for specific gimbal type if one exists (e.g. MAV_COMP_ID_QX1_GIMBAL).</description>
       </entry>
       <entry value="155" name="MAV_COMP_ID_LOG">
-        <description/>
+        <description>Logging component ID.</description>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml]: Are there any particular expectations of a logging component?

File: message_definitions/v1.0/common.xml [hunk 9]
```diff
@@ -369,131 +369,132 @@
...
-        <description>On Screen Display (OSD) devices for video links</description>
+        <description>On Screen Display (OSD) devices for video links.</description>
       </entry>
       <entry value="158" name="MAV_COMP_ID_PERIPHERAL">
-        <description>Generic autopilot peripheral component ID. Meant for devices that do not implement the parameter sub-protocol</description>
+        <description>Generic autopilot peripheral component ID. Meant for devices that do not implement the parameter microservice.</description>
       </entry>
       <entry value="159" name="MAV_COMP_ID_QX1_GIMBAL">
-        <description/>
+        <description>Gimbal ID for QX1.</description>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml]: Why does QX1 need separate addressing?

Comment (julianoes) [reply message_definitions/v1.0/common.xml]: It shouldn't in my opinion.

File: message_definitions/v1.0/common.xml (hunk 7 above)
Comment (julianoes) [reply message_definitions/v1.0/common.xml]: Not sure :smile:

File: message_definitions/v1.0/common.xml (hunk 6 above)
Comment (TSC21) [reply message_definitions/v1.0/common.xml]: @hamishwillee this component is currently set in MAVROS to identify the link between MAVROS and the GCS, which is made through an UDP bridge, forwards the packages between the GCS and the FCU.

**Comment by auturgy** [comment]
Mavesp8266 also identifies as udp bridge (240).

File: message_definitions/v1.0/common.xml [hunk 10]
```diff
@@ -148,7 +148,7 @@
         <description>VTOL reserved 5</description>
       </entry>
       <entry value="26" name="MAV_TYPE_GIMBAL">
-        <description>Onboard gimbal</description>
+        <description>Gimbal (standalone)</description>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml]: FYI, MAV_TYPE is for systems, not component. This was supposed to be fixed in another PR, but will combine it here.

File: message_definitions/v1.0/common.xml (hunk 9 above)
Comment (hamishwillee) [reply message_definitions/v1.0/common.xml]: Thanks. I've marked it as deprecated. FYI @auturgy too. This is clearly wrong and there is a clear right way to do it, so I think deprecation is in order for now, ideally with deletion at some point in future.

File: message_definitions/v1.0/common.xml (hunk 6 above)
Comment (hamishwillee) [reply message_definitions/v1.0/common.xml]: @TSC21 So to clarify, is this a component of the GCS system - so you address the message to GCS and MAV_COMP_ID_UART_BRIDGE, and this tells the GCS to forward the packet to PX4 (say). And the link is via UDP, between MAVROS and QGC (but not necessarily between QGC and PX4)?

Is the MAV_COMP_ID_UART_BRIDGE the same thing?

IN that case, what would you propose for a description, remembering that the user should not have to be ROS?

File: message_definitions/v1.0/common.xml (hunk 7 above)
Comment (hamishwillee) [reply message_definitions/v1.0/common.xml]: Research shows it is used in [AntennaTracker](https://github.com/ArduPilot/ardupilot/blob/master/AntennaTracker/GCS_Mavlink.cpp#L379) to identify the target that should accept an arm command (MAV_CMD_COMPONENT_ARM_DISARM).
Digging a bit further, it seems the requirement to check this component was removed elsewhere from ArduPilot in [2014](https://github.com/ArduPilot/ardupilot/pull/909) and the command will accept MAV_COMP_ID_ALL as a target now. So probably should also be removed from AntennaTracker.

@auturgy I propose to deprecate this (as not used by PX4 and only accidentally by AntennaTracker) and add a description. It would look like the below (OK?)
```
      <entry value="250" name="MAV_COMP_ID_SYSTEM_CONTROL">
        <description>ID for component to handle system commands (e.g. takeoff, arm).</description>
        <deprecated since="2018-11" replaced_by="MAV_COMP_ID_ALL">There is no need to separately address system commands to a specific component for handling.</deprecated>
      </entry>
```

File: message_definitions/v1.0/common.xml (hunk 4 above)
Comment (hamishwillee) [reply message_definitions/v1.0/common.xml]: @mrivi Ping :-)

File: message_definitions/v1.0/common.xml (hunk 2 above)
Comment (hamishwillee) [reply message_definitions/v1.0/common.xml]: @auturgy I can't find this on ArduPilot, PX4 or via google in anything other than header file listings. Other than sending out messages on our forums and google discussions, do you have any ideas? 
Could I add a description "Purpose unknown" and deprecate it after doing that basic due-diligence?

File: message_definitions/v1.0/common.xml (hunk 6 above)
Comment (hamishwillee) [reply message_definitions/v1.0/common.xml]: @auturgy noted that that *Mavesp8266 also identifies as udp bridge (240)*. [Comments](https://github.com/dogmaphobic/mavesp8266/blob/master/src/mavesp8266_component.cpp#L33) say "MavLink UART/UDP Bridge", so essentially this is used for a bridge from UART to UDP. 

WOuld that be an OK description for ROS as well?

File: message_definitions/v1.0/common.xml (hunk 10 above)
Comment (julianoes) [reply message_definitions/v1.0/common.xml]: Yer `MAV_TYPE` is a mess. I think we should probably get rid of it. The only thing that it tells is really the frame type.

Comment (hamishwillee) [reply message_definitions/v1.0/common.xml]: Probably not - this is apparently what is used for determining the icon for airframe in QGC :-)
I think it is fine if used properly.

File: message_definitions/v1.0/common.xml (hunk 2 above)
Comment (hamishwillee) [reply message_definitions/v1.0/common.xml]: @julianoes Any thoughts on this?

File: message_definitions/v1.0/common.xml (hunk 5 above)
Comment (hamishwillee) [reply message_definitions/v1.0/common.xml]: Thanks. Updated!

File: message_definitions/v1.0/common.xml (hunk 6 above)
Comment (hamishwillee) [reply message_definitions/v1.0/common.xml]: FYI, I have updated these. Please check you are happy.

File: message_definitions/v1.0/common.xml [hunk 11]
```diff
@@ -369,131 +369,136 @@
...
       <entry value="159" name="MAV_COMP_ID_QX1_GIMBAL">
-        <description/>
+        <deprecated since="2018-11" replaced_by="MAV_COMP_ID_GIMBAL">All gimbals should use MAV_COMP_ID_GIMBAL.</deprecated>
+        <description>Gimbal ID for QX1.</description>
       </entry>
       <entry value="160" name="MAV_COMP_ID_FLARM">
-        <description/>
+        <description>FLARM collision alert component.</description>
       </entry>
       <entry value="180" name="MAV_COMP_ID_MAPPER">
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml]: Lorenz says this can be deleted.

File: message_definitions/v1.0/common.xml [hunk 12]
```diff
@@ -369,131 +369,136 @@
       </entry>
     </enum>
     <enum name="MAV_COMPONENT">
+      <description>Component ids for the different types and instances of onboard hardware/software that might make up a MAVLink system (autopilot, cameras, servos, GPS systems, avoidance systems etc.). 
+      Components must use the appropriate ID in their source address when sending messages and use the same ID to determine if they are the intended recipient of an incoming message. The MAV_COMP_ID_ALL value is used to indicate messages that must be processed by all components.
+      When creating new entries for this enum, values for components that can have multiple instances (e.g. cameras, servos etc.) should be allocated sequentially (and gaps left in the allocation rage to allow the number of instances to be expanded).</description>
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml]: allocation rage -> allocation range. 
Perhaps provide guidance on what is a sensible range.

File: message_definitions/v1.0/common.xml (hunk 8 above)
Comment (hamishwillee) [reply message_definitions/v1.0/common.xml]: No answer here. But this is IMO sufficient for now.

**Comment by hamishwillee** [comment]
`MAV_COMPONENT` documentation merged as discussed in dev call.

File: message_definitions/v1.0/common.xml (hunk 3 above)
Comment (julianoes) [reply message_definitions/v1.0/common.xml]: @hamishwillee I think your guess about the history of this is correct.

> The messages can be used by multiple systems, and you want to make it clear which system should handle it.

I don't think we need that.

I rather thought that this component ID would be for a component which plans the mission. So let's say this is a companion computer which is responsible for waypoint generation based on some higher level goal.

Comment (hamishwillee) [reply message_definitions/v1.0/common.xml]: Thanks very much! I have clarified this further in https://github.com/mavlink/mavlink/pull/1031
